### ISSUE TITLE
adl2edl crashes on some adl files

### ISSUE THREAD

**Issue by MarkRivers** [epicsdeb/adl2edl#8]
I am getting crashes of adl2edl on a number of adl files.  For example the following files in softGlue:

https://github.com/epics-modules/softGlue/tree/master/softGlueApp/op/adl

- softGlueAll.adl
- softGlueAll_octupole.adl
- softGlueFieldIO_EP202.adl
- softGlueFieldIO_Int_EP202.adl
- softGlueAll_label.adl
- softGlueFieldIO.adl
- softGlueFieldIO_Int.adl

I am running the master branch of adl2edl.  The README says that should be 1.7, but adl2edl -h says it is 1.6.

**Comment by bhill-slac** [comment]
Just ran latest adl2edl on all the files on softGlueApp/op/adl w/ no crashes.
I'm running on rhel6.   Which OS are you running on?

Re version 1.7 I just pushed a fix for that.

**Comment by MarkRivers** [comment]
I tested again on the master branch and it still crashes for me.

I am running on Centos 7, gcc 4.8.5.

This is adl2edl -h to show I am running 1.7
```
corvette:softGlueApp/op/adl>adl2edl -h
Version 1.7
usage: [-h] [-rgb] [-rename] [-d LEVEL] medm_file [edm_file]
   medm_file is the name of an medm file to translate to edm.
      medm_file will typically end in .adl
   edm_file is the edm filename.
      If not used, the translation is written to stdout
      and diagnostic output to ${medm_file}.adl.db
Options:
 -h          show this help message and exit
 -rgb        use rgb mode to specify colors in the edm translation
 -rename     rename the medm file substituting .edl for .adl
 -f fontFile set the font filename
 -d LEVEL    set the diagnostic output to LEVEL
```
This is trying to convert one of the files in softGlue:
```
corvette:softGlueApp/op/adl>adl2edl -f $EDMFONTFILE softGlueAll.adl > softGlueAll.edl
Segmentation fault (core dumped)
```
This is the gdb backtrace of the crash:
```
corvette:softGlueApp/op/adl>gdb adl2edl core.1010
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-110.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /usr/local/adl2edl/adl2edl...done.
[New LWP 1010]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `adl2edl -f /home/epics/devel/configure/fonts.adl2edl softGlueAll.adl'.
Program terminated with signal 11, Segmentation fault.
#0  0x00007f5b89bb12ab in std::string::assign(char const*, unsigned long) () from /lib64/libstdc++.so.6
Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 expat-2.1.0-10.el7_3.x86_64 fontconfig-2.10.95-11.el7.x86_64 freetype-2.8-14.el7.x86_64 glibc-2.17-222.el7.x86_64 libICE-1.0.9-9.el7.x86_64 libSM-1.2.2-2.el7.x86_64 libX11-1.6.5-1.el7.x86_64 libXau-1.0.8-2.1.el7.x86_64 libXext-1.3.3-3.el7.x86_64 libXft-2.3.2-2.el7.x86_64 libXmu-1.1.2-2.el7.x86_64 libXp-1.0.2-2.1.el7.x86_64 libXrender-0.9.10-1.el7.x86_64 libXt-1.1.5-3.el7.x86_64 libgcc-4.8.5-28.el7_5.1.x86_64 libjpeg-turbo-1.2.90-5.el7.x86_64 libpng-1.5.13-7.el7_2.x86_64 libstdc++-4.8.5-28.el7_5.1.x86_64 libuuid-2.23.2-52.el7_5.1.x86_64 libxcb-1.13-1.el7.x86_64 motif-2.3.4-14.el7_5.x86_64 zlib-1.2.7-17.el7.x86_64
(gdb) bt
#0  0x00007f5b89bb12ab in std::string::assign(char const*, unsigned long) () from /lib64/libstdc++.so.6
#1  0x0000000000414ccd in translator::parseComposite (this=0x133e1a0, inf=..., outf=..., outd=..., outt=...) at /usr/local/adl2edl/translator.cc:764
#2  0x00000000004134f9 in translator::processFile (this=0x133e1a0, in=0x7ffdd7f0d09e "softGlueAll.adl", out=0x7ffdd7f0bd90 "softGlueAll.adl.db")
    at /usr/local/adl2edl/translator.cc:491
#3  0x0000000000411a16 in main (argc=4, argv=0x7ffdd7f0bf88) at /usr/local/adl2edl/translator.cc:134
(gdb)
```
Line 764 in translator.cc is this which should not generate an error, so perhaps there is a buffer overflow on the stack, since "title" is an automatic variable?
```
    title = "# (Embedded Window)";
```

**Comment by MarkRivers** [comment]
I found and fixed the problem.  The s1 and s2 buffers in that function were 80 characters, which was not large enough in my case.
```
corvette:/usr/local/adl2edl>git diff
diff --git a/translator.cc b/translator.cc
index 3da0297..92fd86f 100644
--- a/translator.cc
+++ b/translator.cc
@@ -709,7 +709,7 @@ int translator::parseComposite(ifstream &inf, ostream &outf, ostream &outd, ostr
        string squote = "\"";
        string nil("");
     int snum;
-       char s1[80], s2[80];
+       char s1[256], s2[256];
        int x,y,wid,hgt;
        string file;
        int adl_pos;
```
Perhaps a better solution that does not use fixed size buffers should be found.  At the least there should be a warning and protection against buffer overflow.

**Comment by MarkRivers** [comment]
The reason for the buffer overflow were lines like this in the .adl file
```
composite {
	object {
		x=870
		y=780
		width=407
		height=340
	}
	"composite name"=""
	"composite file"="softGlueFieldIOxx_small_bare.adl;P=$(P),H=$(H),N1=33,N2=34,N3=35,N4=36,N5=37,N6=38,N7=39,N8=40,N9=41,N10=42,N11=43,N12=44,N13=45,N14=46,N15=47,N16=48"
}
```
I don't know why that did not also overflow for you?

**Comment by MarkRivers** [comment]
I have another file that crashes adl2edl that occurs even with the above fix.

This is the adl file:
https://github.com/CARS-UChicago/GSECARS-EPICS/blob/master/CARSApp/op/adl/beamQualityReportOps.adl

Here is the crash dump:
```
corvette:CARSApp/op/adl>gdb adl2edl core.26471
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-110.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /usr/local/adl2edl/adl2edl...done.
[New LWP 26471]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `adl2edl -f /home/epics/devel/configure/fonts.adl2edl beamQualityReportOps.adl'.
Program terminated with signal 6, Aborted.
#0  0x00007ff2b94f0277 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 expat-2.1.0-10.el7_3.x86_64 fontconfig-2.10.95-11.el7.x86_64 freetype-2.8-14.el7.x86_64 glibc-2.17-222.el7.x86_64 libICE-1.0.9-9.el7.x86_64 libSM-1.2.2-2.el7.x86_64 libX11-1.6.5-1.el7.x86_64 libXau-1.0.8-2.1.el7.x86_64 libXext-1.3.3-3.el7.x86_64 libXft-2.3.2-2.el7.x86_64 libXmu-1.1.2-2.el7.x86_64 libXp-1.0.2-2.1.el7.x86_64 libXrender-0.9.10-1.el7.x86_64 libXt-1.1.5-3.el7.x86_64 libgcc-4.8.5-28.el7_5.1.x86_64 libjpeg-turbo-1.2.90-5.el7.x86_64 libpng-1.5.13-7.el7_2.x86_64 libstdc++-4.8.5-28.el7_5.1.x86_64 libuuid-2.23.2-52.el7_5.1.x86_64 libxcb-1.13-1.el7.x86_64 motif-2.3.4-14.el7_5.x86_64 zlib-1.2.7-17.el7.x86_64
(gdb) bt
#0  0x00007ff2b94f0277 in raise () from /lib64/libc.so.6
#1  0x00007ff2b94f1968 in abort () from /lib64/libc.so.6
#2  0x00007ff2b9532d37 in __libc_message () from /lib64/libc.so.6
#3  0x00007ff2b953b499 in _int_free () from /lib64/libc.so.6
#4  0x00007ff2b9e5cc03 in std::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() () from /lib64/libstdc++.so.6
#5  0x0000000000422094 in shellnode::~shellnode (this=0x243db30, __in_chrg=<optimized out>) at /usr/local/adl2edl/control.cc:386
#6  0x000000000042813d in std::_Destroy<shellnode> (__pointer=0x243db30) at /usr/include/c++/4.8.2/bits/stl_construct.h:93
#7  0x0000000000427db2 in std::_Destroy_aux<false>::__destroy<shellnode*> (__first=0x243db30, __last=0x243db50)
    at /usr/include/c++/4.8.2/bits/stl_construct.h:103
#8  0x0000000000427723 in std::_Destroy<shellnode*> (__first=0x243db30, __last=0x243db50) at /usr/include/c++/4.8.2/bits/stl_construct.h:126
#9  0x0000000000426bf9 in std::_Destroy<shellnode*, shellnode> (__first=0x243db30, __last=0x243db50) at /usr/include/c++/4.8.2/bits/stl_construct.h:151
#10 0x00000000004268ce in std::vector<shellnode, std::allocator<shellnode> >::~vector (this=0x7ffcc9b483d0, __in_chrg=<optimized out>)
    at /usr/include/c++/4.8.2/bits/stl_vector.h:415
#11 0x0000000000422f07 in shellclass::parse (this=0x7ffcc9b48b80, inf=..., outf=..., outd=...) at /usr/local/adl2edl/control.cc:507
#12 0x0000000000412d16 in translator::processFile (this=0x24351a0, in=0x7ffcc9b4b09d "beamQualityReportOps.adl",
    out=0x7ffcc9b49000 "beamQualityReportOps.adl.db") at /usr/local/adl2edl/translator.cc:374
#13 0x0000000000411a16 in main (argc=4, argv=0x7ffcc9b491f8) at /usr/local/adl2edl/translator.cc:134
(gdb)
```

**Comment by mdavidsaver** [comment]
It might be worthwhile to build with `-fstack-protector-strong`.  Not as a fix, but it should help localize problems faster.  https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#Instrumentation-Options

The current Debian recommendations are listed (and explained/justified) at https://wiki.debian.org/Hardening

Compile: `-Wdate-time -D_FORTIFY_SOURCE=2 -fstack-protector-strong -Wformat -Werror=format-security`
Link: `-Wl,-z,relro`

**Comment by MarkRivers** [comment]
I tried the compile options (-D_FORTIFY_SOURCE=2 -O -fstack-protector-strong -Wformat -Werror=format-security) and it did not help.  -Wdate-time was not allowed on my compiler.

**Comment by bhill-slac** [comment]
Using gcc 4.4.7 so best I could do was add -fstack-protector-all.   Still doesn't crash but I agree all those sscanf calls w/ %s are a problem.    Don't have time to fix now so I just bumped up all those
static s1 and s2 buffers to 256.

The other problem occurs because adl format allows a label for both the shell
command and it's widget.  The widget label isn't supported by edl
and was being parsed outside the shellNode context.

I added guards for shellCommand and relatedClass parsing and created new 1.8 release.
Let me know if it works for you.
